### PR TITLE
Fix witness generator for failed call/create op and tx

### LINKED ISSUE #145: Missing trace for caller in "CREATE" op
Currently the `extraData` field for `CRAETE` op contains only states for the account of created contract. However, the caller's nonce is also increased and the balance may be updated.

We can capture these updating from the `To` field if the contract called by user has executed an creation. But we would definitely lost some states for the deeper creation (user call contract A, A called B, then B create C). We must capture the caller's state in `extraData` field.

I advise this fixing should be made together with issue #144 because they are updates `before CaptureStart`. Also, we must notice there are in fact two diferent kinds of updating: the increase on caller's nonce is not snapshot (i.e. it would not reverted even the op has failed) while the updating on balance is revertable.

### PR DESCRIPTION
**Pull request by noel2004**
Current witness generator just skip every entries in execution for a failed tx, which is never consistent with the expection of state circuit: though not affect the final state, we had to provide some "read" action inside a failed execution sequences.

A "readonly" mode is added to the witness generator and under this mode all accessing is recorded and reverted when the mode is disabled, just like the snapshot restortion in evm.

Currently we still need some issue (#144, #145) being resolved to made the reverting function in every case, and more strait tests may be required for some edge cases like:

- [x]  A failed call/creat with recursive call/creats: contract A call B, and B call/create C, success or failed, then the call of B reverted.
- [x] A failed call with value reverted by 'no enough balance'
- [x] The nonce updating on a deep, recursive call of `CREATE` (contract A call B, B create C, nonce of B must be updated)
- [x] `DELEGATECALL`

Notes:

> 1. By inducing a capture of "caller" status in trace fields we have resolved https://github.com/scroll-tech/go-ethereum/issues/145 now.
> 
>     Now this PR would ensure correction on many edge cases but still no hope to be 100% consistent with the entries in mpt_table for some immediate failure (i.e. failed by call with no enough balance or create with address collision). The later has been planned to be resolved in a long-term project.
> 
> 2. It should also be noted that with this PR witness generator can handle the case of account deletion (i.e. SELFDESTRUCT op) naturally. For it have to handle the reverting of account creation, which is equal to an account deletion

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1202799565449149

### CODE REVIEW

**Comment by noel2004** [comment]
By inducing a capture of "caller" status in trace fields we have resolved #145 now.

Now this PR would ensure correction on many edge cases but still no hope to be 100% consistent with the entries in mpt_table for some immediate failure (i.e. failed by call with no enough balance or create with address collision). The later has been planned to be resolved in a long-term project.

**Comment by noel2004** [comment]
Now this PR has also included most works in PR #140: handle the selfdestruct case (also not consistent with mpt table)

**Comment by lispc** [comment]
After https://github.com/scroll-tech/go-ethereum/pull/147,  i think we should write self-destruct related codes like this

We may re-consider the "treat selfdestruct as invalid" decision later. But in mid term, self destruct will be treated as invalid.

```
var selfDestructAsInvalid = true 

...
if selfDestructAsInvalid {
 ..??
} else {
 // the codes in this pr
}
...

```

File: trie/zkproof/writer.go [hunk 1]
```diff
@@ -263,7 +264,16 @@ func copyAccountState(st *types.AccountWrapper) *types.AccountWrapper {
 	}
 }
 
+func isDeletedAccount(state *types.AccountWrapper) bool {
```
Comment (mask-pp) [inline trie/zkproof/writer.go:272]: How about store `xx_trace.json` into internal module?

Comment (noel2004) [reply trie/zkproof/writer.go:272]: Those files are just stuff for unit testing, it may be too wired to separate them from the test itself ... so you mean we should put the test in some path inside internal? or they would be shared as stuff for some other tests?

Comment (mask-pp) [reply trie/zkproof/writer.go:272]: > 

Yes, i can also use that data in some test cases when add compress logic for `BlockResult struct`.

Comment (noel2004) [reply trie/zkproof/writer.go:272]: ok, for which path they are suitable? what about `internal/utesting/blocktraces` ?

**Comment by lispc** [comment]
@noel2004  do you think we need / or better rengerate all the trace jsons in common-rs after this PR?

need not? since it is related to failed tx

File: trie/zkproof/writer.go (hunk 1 above)
Comment (mask-pp) [reply trie/zkproof/writer.go:272]: ok, thanks.

**Comment by noel2004** [comment]
> @noel2004 do you think we need / or better rengerate all the trace jsons in common-rs after this PR?
> 
> need not? since it is related to failed tx

We should not need that. If `mptwitness` field has been generated in the trace jsons, they would be extractly the same as which being generated under curerent PR.

**Comment by noel2004** [comment]
It should also be noted that with this PR witness generator can handle the case of account deletion (i.e. `SELFDESTRUCT` op) naturally. For it have to handle the reverting of account creation, which is equal to an account deletion

**Comment by lispc** [comment]
lgtm. the only concern is "self destruct" related changes. @Thegaram can have a look

**Review by lispc — DISMISSED** [review]

File: trie/zkproof/writer.go (hunk 1 above)
Comment (noel2004) [reply trie/zkproof/writer.go:272]: done

**Review by lispc — DISMISSED** [review]
only one todo later: 

1. add a trace json with self-destruct to common rs. We have to test this case.

**Comment by noel2004** [comment]
> only one todo later:
> 
> 1. add a trace json with self-destruct to common rs. We have to test this case.

Yes. I found we need to update some code since SELFDESTRUCT has been disable.

I have update the selfdestruct trace file, which contain 2 txs with [some rather edge behaviors](https://github.com/scroll-tech/mpt-circuit/blob/master/integration-tests/contracts/Destruct.sol) and it can be put in common-rs

**Comment by lispc** [comment]
> I found we need to update some code since SELFDESTRUCT has been disable.

yes that is what i mean in https://github.com/scroll-tech/go-ethereum/pull/146#issuecomment-1225082590

**Comment by noel2004** [comment]
>

**Comment by noel2004** [comment]
> > I found we need to update some code since SELFDESTRUCT has been disable.
> 
> yes that is what i mean in [#146 (comment)](https://github.com/scroll-tech/go-ethereum/pull/146#issuecomment-1225082590)

What about just handle the disable case of self-destruct in this PR? the reason:

1. #147 simple disable self-destruct without any options to re-enable it. It may take another full PR to add the flags and codes to control the behavior by a switch in run time
2.  Witness generator itself in zkgeth may be replaced by which in common-rs, I still not sure which would be first (switchable enabling of self-destruct or porting mpt witness generating to roller)

**Comment by lispc** [comment]
What about just handle the disable case of self-destruct in this PR: no problem

**Comment by 0xmountaintop** [comment]
I think we can rename "XXX_trace.json" to "XXX.json" coz they are already inside "blocktraces/" folder

**Comment by noel2004** [comment]
> I think we can rename "XXX_trace.json" to "XXX.json" coz they are already inside "blocktraces/" folder

Ok

**Review by mask-pp — DISMISSED** [review]

**Comment by 0xmountaintop** [comment]
do we need to regenerate the traces after https://github.com/scroll-tech/go-ethereum/pull/149 https://github.com/scroll-tech/go-ethereum/pull/157 https://github.com/scroll-tech/go-ethereum/pull/158?

**Comment by noel2004** [comment]
> do we need to regenerate the traces after #149 #157 #158?

For testing witness generator only, it is not. For other purpose, maybe. 

Maybe we have considered update / move them into separated repo so in this PR can we just leave the traces as is ?

**Review by 0xmountaintop — APPROVED** [review]

**Review by scroll-dev — APPROVED** [review]
